server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,226 @@
+        clusterService.addListener(this);
+    }
+
+    // package private for testing
+    boolean ready() {
+        return this.serverChannel != null;
+    }
+
+    /**
+     * Checks to see if the readiness service is enabled in the current environment
+     * @param environment
+     * @return
+     */
+    public static boolean enabled(Environment environment) {
+        return PORT.get(environment.settings()) != -1;
+    }
+
+    // package private for testing
+    ServerSocketChannel serverChannel() {
+        return serverChannel;
+    }
+
+    /**
+     * Returns the current bound address for the readiness service.
+     * If Elasticsearch was never ready, this method will return null.
+     * @return the bound address for the readiness service
+     */
+    public BoundTransportAddress boundAddress() {
+        InetSocketAddress boundAddress = boundSocket.get();
+        if (boundAddress == null) {
+            return null;
+        }
+        TransportAddress publishAddress = new TransportAddress(boundAddress);
+        return new BoundTransportAddress(new TransportAddress[] { publishAddress }, publishAddress);
+    }
+
+    ServerSocketChannel setupSocket() {
+        assert PORT.get(environment.settings()) >= 0;
+        InetAddress localhost = InetAddress.getLoopbackAddress();
+        int portNumber = PORT.get(environment.settings());
+
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            try {
+                serverChannel = ServerSocketChannel.open(StandardProtocolFamily.INET);
+                if (boundSocket.get() != null) {
+                    serverChannel.bind(boundSocket.get());
+                } else {
+                    InetSocketAddress socketAddress = new InetSocketAddress(localhost, portNumber);
+                    serverChannel.bind(socketAddress);
+
+                    boundSocket.set((InetSocketAddress) serverChannel.getLocalAddress());
+
+                    BoundTransportAddress boundAddress = boundAddress();
+                    for (BoundAddressListener listener : boundAddressListeners) {
+                        listener.addressBound(boundAddress);
+                    }
+                }
+            } catch (Exception e) {
+                throw new BindTransportException("Failed to bind to " + NetworkAddress.format(localhost, portNumber), e);
+            }
+
+            return null;
+        });
+
+        return serverChannel;
+    }
+
+    @Override
+    protected void doStart() {
+        // Mark the service as active, we'll start the listener when ES is ready
+        this.active = true;
+    }
+
+    // package private for testing
+    synchronized void startListener() {
+        assert enabled(environment);
+
+        if (this.serverChannel != null || this.active == false) {
+            return;
+        }
+
+        this.serverChannel = setupSocket();
+        this.listenerThreadLatch = new CountDownLatch(1);
+
+        new Thread(() -> {
```

Comment:
Optionally, we could choose to create this thread with a very small stack size, rather than the default size. (since the thread will use very little stack)